src/main/java/org/zalando/nakadi/config/NakadiConfig.java

```diff
@@ -34,22 +39,54 @@ public ZooKeeperLockFactory zooKeeperLockFactory(final ZooKeeperHolder zooKeeper
 
     @Bean
     @Qualifier("default_storage")
-    public Storage defaultStorage(final StorageDbRepository storageDbRepository,
-                                  final Environment environment) throws InternalNakadiException {
-        final Storage storage = new Storage();
-        storage.setId("default");
-        storage.setType(Storage.Type.KAFKA);
-        storage.setConfiguration(new Storage.KafkaConfiguration(
-                environment.getProperty("nakadi.zookeeper.exhibitor.brokers"),
-                Integer.valueOf(environment.getProperty("nakadi.zookeeper.exhibitor.port", "0")),
-                environment.getProperty("nakadi.zookeeper.brokers"),
-                environment.getProperty("nakadi.zookeeper.kafkaNamespace", "")));
+    public DefaultStorage defaultStorage(final StorageDbRepository storageDbRepository,
+                                         final Environment environment,
+                                         final ZooKeeperHolder zooKeeperHolder)
+            throws InternalNakadiException {
+        final String storageId = getStorageId(zooKeeperHolder, environment);
+        final Optional<Storage> storageOpt = storageDbRepository.getStorage(storageId);
+        if (!storageOpt.isPresent()) {
+            LOGGER.info("Creating timelines storage `{}` from defaults", storageId);
+            final Storage storage = new Storage();
+            storage.setId(storageId);
+            storage.setType(Storage.Type.KAFKA);
+            storage.setConfiguration(new Storage.KafkaConfiguration(
+                    environment.getProperty("nakadi.zookeeper.exhibitor.brokers"),
+                    Integer.valueOf(environment.getProperty("nakadi.zookeeper.exhibitor.port", "0")),
+                    environment.getProperty("nakadi.zookeeper.brokers"),
+                    environment.getProperty("nakadi.zookeeper.kafkaNamespace", "")));
+            try {
+                storageDbRepository.createStorage(storage);
+            } catch (final DuplicatedStorageException e) {
+                LOGGER.info("Creation of default storage failed: {}", e.getMessage());
+            }
+            return new DefaultStorage(storage);
+        } else {
+            return new DefaultStorage(storageOpt.get());
+        }
+    }
+
+    private String getStorageId(final ZooKeeperHolder zooKeeperHolder,
+                                final Environment environment) {
+        final CuratorFramework curator = zooKeeperHolder.get();
         try {
-            storageDbRepository.createStorage(storage);
-        } catch (final DuplicatedStorageException e) {
-            LOGGER.info("Creation of default storage failed: {}", e.getMessage());
+            curator.create().creatingParentsIfNeeded()
+                    .forPath(StorageService.ZK_TIMELINES_DEFAULT_STORAGE, null);
+        } catch (final Exception e) {
```

Comment:
I think you should catch more specific exception here (I think its NodeExistsException or something like that).

Reply:
Basically, I do not care what happens on that step, if it fails I will take from env var. I can fix the message here

Reply:
With current implementation each time Nakadi instance starts it will log ERROR in the logs. I think it's better to make difference between NodeExists exception and other exceptions.